### PR TITLE
scala-next: Init at 3.5.1

### PR DESCRIPTION
**Pull request by nikolaiser**
## Description of changes
This change allows to use the latest non-LTS scala version which is currently 3.5.1. Since 3.5.1 now ships with the `scala-cli` I needed to modify the `preFixup` attribute.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by kashw2 — COMMENTED** [review]
Hi and Thanks and congratulations on your first contribution to Nixpkgs.

Looking over this, I'm not sure we need to provide a package for each minor version of Scala 3.

Generally I see this reserved for cases where we may need to override specific package's versions of a runtime/compiler to achieve a building or running package, an example of this is the go derivations and their usage in all-packages.nix. Also to note, the usage of go throughout Nixpkgs is extensive to the point that just overriding in go packages themselves would constitute a lot of reused code.

I'll defer to other maintainers for their opinions on something like this.

**Comment by nikolaiser** [comment]
@kashw2 thank you for your feedback! Instead of defining packages for every minor version would you then be open to having a new package called `scala-next`that always points to the latest non-LTS version?

File: pkgs/development/compilers/scala/bare.nix
Comment (nikolaiser) [reply pkgs/development/compilers/scala/bare.nix]: Fixed

**Comment by SuperSandro2000** [comment]
Yeah, only adding the latest version next to the lts version sounds fine to me.

**Comment by nikolaiser** [comment]
I removed all new packages except for 3.5.0 which is now named `scala-next`

**Comment by karolchmist** [comment]
Result of `nixpkgs-review pr 336735` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>scala (scala_3)</li>
    <li>scala-next</li>
    <li>tlaps</li>
  </ul>
</details>

**Review by karolchmist — APPROVED** [review]

**Review by kashw2 — APPROVED** [review]

**Review by s0me1newithhand7s — APPROVED** [review]
LGTM

File: pkgs/development/compilers/scala/bare.nix
Comment (natsukagami) [inline pkgs/development/compilers/scala/bare.nix:45]: I think we should no longer advertise Dotty as an experimental platform, as it is now *the* Scala 3 compiler. Simplifying it to just

> The Scala 3 compiler, also known as Dotty.

Should work imo.

**Review by natsukagami — COMMENTED** [review]
I (from Scala Center) would very appreciate if you would take some tweaks I've made to the package here: https://github.com/nikolaiser/nixpkgs/pull/1
I was going to make a separate PR, but I think it's just better if we tweaked this already awesomely made branch a little bit :) Thanks!

**Comment by nikolaiser** [comment]
@natsukagami Thanks for your PR!

**Review by s0me1newithhand7s — CHANGES_REQUESTED** [review]
you need to have 1 commit per file. for now we have: 4 files (including `maintainers-list.nix`) and 9 commits.

**Comment by nikolaiser** [comment]
Thanks for the pointer. I squashed it down to 2 commits

I also updated the `scala-next` package to 3.5.1 since it was released in the meantime

**Review by natsukagami — APPROVED** [review]

**Comment by natsukagami** [comment]
@kubukoz would be glad to get a review/get it merged :)

File: pkgs/development/compilers/scala/bare.nix
Comment (kubukoz) [inline pkgs/development/compilers/scala/bare.nix:10]: TIL - I didn't know you could do this in `mkDerivation`

**Comment by kubukoz** [comment]
Result of `nixpkgs-review pr 336735` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>isabelle</li>
    <li>isabelle-components.isabelle-linter</li>
    <li>tlaps</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>scala (scala_3)</li>
    <li>scala-next</li>
  </ul>
</details>

**Review by kubukoz — APPROVED** [review]

**Comment by natsukagami** [comment]
What else do we need for it to be merged?

**Comment by kubukoz** [comment]
I was just waiting for CI and didn't come back until now 😅

**Pull request merged**
